File: src/tools/farchc.c
/***************************************************************************/
/*                                                                         */
/* Copyright 2022 INTERSEC SA                                              */
/*                                                                         */
/* Licensed under the Apache License, Version 2.0 (the "License");         */
/* you may not use this file except in compliance with the License.        */
/* You may obtain a copy of the License at                                 */
/*                                                                         */
/*     http://www.apache.org/licenses/LICENSE-2.0                          */
/*                                                                         */
/* Unless required by applicable law or agreed to in writing, software     */
/* distributed under the License is distributed on an "AS IS" BASIS,       */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.*/
/* See the License for the specific language governing permissions and     */
/* limitations under the License.                                          */
/*                                                                         */
/***************************************************************************/

#include <lib-common/parseopt.h>
#include <lib-common/qlzo.h>
#include <lib-common/unix.h>
#include <lib-common/farch.h>

static struct opts_g {
    const char *out;
    const char *target;
    const char *deps;
    int help;
    int verbose;
    int compress_lzo;
} opts_g;

static popt_t popt[] = {
    OPT_FLAG('h', "help",         &opts_g.help,    "show help"),
    OPT_FLAG('v', "verbose",      &opts_g.verbose, "be verbose"),
    OPT_STR('d',  "deps",         &opts_g.deps,    "build depends file"),
    OPT_STR('o',  "output",       &opts_g.out,
            "output to this file, default: stdout"),
    OPT_STR('T',  "target",       &opts_g.target,
            "add that to the dep target"),
    OPT_FLAG('c', "compress-lzo", &opts_g.compress_lzo,
             "compress files using LZO algorithm"),
    OPT_END(),
};

static void panic_purge(void)
{
    if (opts_g.out) {
        unlink(opts_g.out);
    }
    exit(EXIT_FAILURE);
}

#define TRACE(fmt, ...) \
    do { if (opts_g.verbose) e_info("farchc: "fmt, ##__VA_ARGS__); } while (0)

#define DIE_IF(tst, fmt, ...) \
    do { if (tst) { e_error(fmt, ##__VA_ARGS__); panic_purge(); } } while (0)

static void put_as_str(lstr_t chunk, FILE *out)
{
    for (int i = 0; i < chunk.len; i++) {
        fprintf(out, "\\x%x", chunk.s[i]);
    }
}

static void put_chunk(lstr_t chunk, FILE *out)
{
    fprintf(out,
            "    LSTR_IMMED(\"");
    put_as_str(chunk, out);
    fprintf(out, "\"),\n");
}

static int dump_and_obfuscate(const char *data, int len, FILE *out)
{
    int nb_chunk = 0;

    while (len > 0) {
        char buffer[FARCH_MAX_SYMBOL_SIZE];
        lstr_t obfuscated_chunk;
        uint32_t chunk_size;

        chunk_size = rand_range(FARCH_MAX_SYMBOL_SIZE / 2,
                                FARCH_MAX_SYMBOL_SIZE);
        chunk_size = MIN((uint32_t)len, chunk_size);
        obfuscated_chunk = LSTR_INIT_V(buffer, chunk_size);
        lstr_obfuscate(LSTR_INIT_V(data, chunk_size), chunk_size,
                       obfuscated_chunk);
        put_chunk(obfuscated_chunk, out);
        data += chunk_size;
        len -= chunk_size;
        nb_chunk++;
    }

    assert (len == 0);

    return nb_chunk;
}

static void dump_file(const char *path, farch_entry_t *entry, FILE *out)
{
    lstr_t file;

    DIE_IF(lstr_init_from_file(&file, path, PROT_READ, MAP_SHARED) < 0,
           "unable to open `%s` for reading: %m", path);
    entry->size = file.len;

    if (opts_g.compress_lzo) {
        t_scope;
        byte lzo_buf[LZO_BUF_MEM_SIZE];
        char *cbuf;
        int clen;

        clen = lzo_cbuf_size(file.len);
        cbuf = t_new_raw(char, clen);
        clen = qlzo1x_compress(cbuf, clen, ps_initlstr(&file), lzo_buf);

        if (clen < file.len) {
            entry->nb_chunks = dump_and_obfuscate(cbuf, clen, out);
            entry->compressed_size = clen;
        } else {
            entry->nb_chunks = dump_and_obfuscate(file.s, file.len, out);
            entry->compressed_size = file.len;
        }
    } else {
        entry->nb_chunks = dump_and_obfuscate(file.s, file.len, out);
        entry->compressed_size = file.len;
    }

    lstr_wipe(&file);
}

qvector_t(farch_entry, farch_entry_t);

static void dump_entries(const char *archname,
                         const qv_t(farch_entry) *entries, FILE *out)
{
    int chunk = 0;

    tab_for_each_ptr(entry, entries) {
        char buffer[2 * PATH_MAX];
        lstr_t obfuscated_name = LSTR_INIT(buffer, entry->name.len);

        lstr_obfuscate(entry->name, entry->nb_chunks, obfuscated_name);

        fprintf(out, "/* {""{{ %*pM */\n", LSTR_FMT_ARG(entry->name));
        fprintf(out, "{\n"
                "    .name = LSTR_IMMED(\"");
        put_as_str(obfuscated_name, out);
        fprintf(out, "\"),\n"
                "    .chunks = &%s_data[%d],\n"
                "    .size = %d,\n"
                "    .compressed_size = %d,\n"
                "    .nb_chunks = %d,\n"
                "},\n"
                "/* }""}} */\n",
                archname, chunk, entry->size, entry->compressed_size,
                entry->nb_chunks);
        chunk += entry->nb_chunks;
    }
}

static int do_work(const char *reldir, FILE *in, FILE *out, FILE *deps)
{
    t_scope;
    SB_1k(dep);
    char srcdir[PATH_MAX];
    char buf[PATH_MAX];
    char name[PATH_MAX];
    int srcdirlen = 0;
    qv_t(farch_entry) entries;

    t_qv_init(&entries, 10);

    do {
        DIE_IF(!fgets(name, sizeof(name), in), "no variable name specified: %m");
        DIE_IF(name[strlen(name) - 1] != '\n', "line 1 is too long: %m");
    } while (*skipspaces(name) == '#' || *skipspaces(name) == '\0');
    strrtrim(name);
    TRACE("creating `%s`", name);

    fprintf(out, "/* This file is generated by farchc. */\n");
    fprintf(out, "\n");
    fprintf(out, "#include <lib-common/farch.h>\n");
    fprintf(out, "\n");
    fprintf(out, "static const farch_data_t %s_data[] = {\n", name);

    if (opts_g.target)
        sb_addf(&dep, "%s%s ", reldir, opts_g.target);
    if (opts_g.out)
        sb_addf(&dep, "%s ", opts_g.out);
    sb_addf(&dep, "%s: ", opts_g.deps);

    snprintf(srcdir, sizeof(srcdir), "%s", reldir);
    path_simplify(srcdir);
    path_join(srcdir, sizeof(srcdir), "/");
    srcdirlen = strlen(srcdir);

    for (int lineno = 2; fgets(buf, sizeof(buf), in); lineno++) {
        const char *s = skipspaces(buf);
        char path[PATH_MAX];
        farch_entry_t entry;
        char *fullname;

        DIE_IF(buf[strlen(buf) - 1] != '\n', "line %d is too long", lineno);
        strrtrim(buf);
        if (!*s) {
            continue;
        }
        if (*s == '#') {
            TRACE("%s", s);
            continue;
        }

        snprintf(path, sizeof(path), "%s%s", srcdir, s);
        fullname = path + srcdirlen;

        if (deps) {
            fprintf(deps, "%s%s\n", dep.data, path);
            fprintf(deps, "%s:\n", path);
        }

        TRACE("adding `%s` as `%s`", path, fullname);
        entry.name = t_lstr_dups(fullname, -1);
        fprintf(out, "/* {""{{ %s */\n", fullname);
        dump_file(path, &entry, out);
        fprintf(out, "/* }""}} */\n");
        qv_append(&entries, entry);
    }

    fprintf(out, "};\n\n"
            "static const farch_entry_t %s[] = {\n", name);
    dump_entries(name, &entries, out);

    fprintf(out,
            "{   .name = LSTR_NULL },\n"
            "};\n");
    return 0;
}

int main(int argc, char *argv[])
{
    t_scope;
    const char *arg0 = NEXTARG(argc, argv);
    char reldir[PATH_MAX];
    FILE *in = stdin, *out = stdout, *deps = NULL;

    argc = parseopt(argc, argv, popt, 0);
    if (argc < 0 || argc > 1 || opts_g.help)
        makeusage(EXIT_FAILURE, arg0, "<farch-script>", NULL, popt);

    if (opts_g.out) {
        out = fopen(opts_g.out, "w");
        DIE_IF(!out, "unable to open `%s` for writing: %m", opts_g.out);
    }

    if (argc > 0) {
        in = fopen(argv[0], "r");
        path_dirname(reldir, sizeof(reldir), argv[0]);
        path_join(reldir, sizeof(reldir), "/");
        DIE_IF(!in, "unable to open `%s` for reading: %m", argv[0]);
    } else {
        reldir[0] = '\0';
    }

    if (opts_g.deps) {
        deps = fopen(opts_g.deps, "w");
        DIE_IF(!deps, "unable to open `%s` for writing: %m", opts_g.deps);
    }

    do_work(reldir, in, out, deps);
    p_fclose(&in);
    p_fclose(&out);
    p_fclose(&deps);
    if (opts_g.out) {
        DIE_IF(chmod(opts_g.out, 0440), "unable to chmod `%s`: %m",
               opts_g.out);
    }
    TRACE("OK !");
    return 0;
}

File: rust/farchc/farchc.rs
/***************************************************************************/
/*                                                                         */
/* Copyright 2025 INTERSEC SA                                              */
/*                                                                         */
/* Licensed under the Apache License, Version 2.0 (the "License");         */
/* you may not use this file except in compliance with the License.        */
/* You may obtain a copy of the License at                                 */
/*                                                                         */
/*     http://www.apache.org/licenses/LICENSE-2.0                          */
/*                                                                         */
/* Unless required by applicable law or agreed to in writing, software     */
/* distributed under the License is distributed on an "AS IS" BASIS,       */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.*/
/* See the License for the specific language governing permissions and     */
/* limitations under the License.                                          */
/*                                                                         */
/***************************************************************************/

use std::fmt::Write;
use std::fs::{self, File};
use std::io::{self, Write as _};
use std::os::unix::fs::PermissionsExt;
use std::path::{Path, PathBuf};

use clap::Parser;

mod bindings {
    #![allow(non_upper_case_globals)]
    #![allow(non_camel_case_types)]
    #![allow(non_snake_case)]
    include!("_bindings.rs");
}
use bindings::*;

impl From<&[u8]> for lstr_t {
    fn from(buf: &[u8]) -> Self {
        Self {
            __bindgen_anon_1: lstr_t__bindgen_ty_1 {
                s: buf.as_ptr() as *const i8,
            },
            len: buf.len() as i32,
            mem_pool: 0,
        }
    }
}

const PATHMAX: i32 = 4096;
const LZO_BUF_MEM_SIZE: usize = 1 << (14 + std::mem::size_of::<u32>());
const FARCH_MAX_SYMBOL_SIZE: usize = 128;

struct FarchEntry {
    pub name: String,
    pub compressed_size: i32,
    pub size: i32,
    pub nb_chunks: i32,
}

/// Processes a farch script
#[derive(Debug, Parser)]
#[command(name = "farch", about = "Processes a farch script")]
struct Opts {
    /// The farch script file to process
    script: PathBuf,

    /// Output file
    #[arg(short, long, value_name = "FILE")]
    out: Option<PathBuf>,

    /// Be verbose
    #[arg(short, long, action = clap::ArgAction::SetTrue)]
    verbose: bool,
}

fn rand_range(first: usize, last: usize) -> usize {
    rand::random_range(first..=last)
}

fn put_as_str(data: &[u8], output: &mut dyn Write) {
    for byte in data {
        write!(output, "\\x{:x}", byte).unwrap();
    }
}

fn put_chunk(chunk: &[u8], output: &mut dyn Write) {
    write!(output, "    LSTR_IMMED(\"").unwrap();
    put_as_str(chunk, output);
    writeln!(output, "\"),").unwrap();
}

fn obfuscate_data(data: &[u8], key: u64, output: &[u8]) {
    let src: lstr_t = data.into();
    let dst: lstr_t = output.into();

    unsafe {
        lstr_obfuscate(src, key, dst);
    }
}

fn dump_and_obfuscate(buf: &[u8], output: &mut dyn Write) -> i32 {
    let mut nb_chunks: i32 = 0;
    let mut len = buf.len();
    let mut start_slice: usize = 0;

    while len > 0 {
        let chunk_size = rand_range(FARCH_MAX_SYMBOL_SIZE / 2, FARCH_MAX_SYMBOL_SIZE);
        let chunk_size = std::cmp::min(chunk_size, len);
        let end_slice: usize = start_slice + chunk_size;
        let obfuscated_chunk_slice = &buf[start_slice..end_slice];
        let obfuscated_output_buf = [0u8; FARCH_MAX_SYMBOL_SIZE];
        let obfuscated_output = &obfuscated_output_buf[0..chunk_size];

        obfuscate_data(obfuscated_chunk_slice, chunk_size as u64, obfuscated_output);
        put_chunk(obfuscated_output, output);

        start_slice = end_slice;
        len -= chunk_size;
        nb_chunks += 1;
    }

    nb_chunks
}

fn dump_file(path: &Path, entry: &mut FarchEntry, output: &mut dyn Write) {
    let file_data = std::fs::read(path).unwrap_or_else(|e| {
        eprintln!("Error: unable to read file `{}`: {}", path.display(), e);
        std::process::exit(1);
    });

    entry.size = file_data.len() as i32;

    let mut clen = unsafe { lzo_cbuf_size(file_data.len()) };
    let mut cbuf = vec![0u8; clen];
    let mut lzo_buf = [0u8; LZO_BUF_MEM_SIZE];
    let start_ptr: *const u8 = file_data.as_ptr();
    let end_ptr: *const u8 = file_data.as_slice().last().unwrap();
    let ps = unsafe {
        pstream_t {
            __bindgen_anon_1: pstream_t__bindgen_ty_1 { b: start_ptr },
            __bindgen_anon_2: pstream_t__bindgen_ty_2 {
                b_end: end_ptr.offset(1),
            },
        }
    };

    unsafe {
        clen = qlzo1x_compress(
            cbuf.as_mut_ptr() as *mut std::os::raw::c_void,
            clen,
            ps,
            lzo_buf.as_mut_ptr() as *mut std::os::raw::c_void,
        );
    }

    if clen < file_data.len() {
        entry.nb_chunks = dump_and_obfuscate(&cbuf[0..clen], output);
        entry.compressed_size = clen as i32;
    } else {
        entry.nb_chunks = dump_and_obfuscate(&file_data, output);
        entry.compressed_size = file_data.len() as i32;
    }
}

fn dump_entries(archname: &str, entries: &[FarchEntry], output: &mut dyn Write) {
    let mut chunk = 0;

    for entry in entries {
        let obfuscated_name = vec![0u8; entry.name.len()];

        obfuscate_data(
            entry.name.as_bytes(),
            entry.nb_chunks as u64,
            &obfuscated_name,
        );
        writeln!(output, "/* {{{{{{ {} */", entry.name).unwrap();
        writeln!(output, "{{").unwrap();
        write!(output, "    .name = LSTR_IMMED(\"").unwrap();
        put_as_str(&obfuscated_name, output);
        writeln!(output, "\"),").unwrap();
        writeln!(output, "    .chunks = &{}_data[{}],", archname, chunk).unwrap();
        writeln!(output, "    .size = {},", entry.size).unwrap();
        writeln!(output, "    .compressed_size = {},", entry.compressed_size).unwrap();
        writeln!(output, "    .nb_chunks = {},", entry.nb_chunks).unwrap();
        writeln!(output, "}},").unwrap();
        writeln!(output, "/* }}}}}} */").unwrap();
        chunk += entry.nb_chunks;
    }
}

fn do_work(opts: &Opts, reldir: &Path, output: &mut impl Write) {
    let content = std::fs::read_to_string(&opts.script).unwrap_or_else(|e| {
        eprintln!(
            "Error: could not read script file `{}`: {}",
            opts.script.display(),
            e
        );
        std::process::exit(1);
    });

    let mut lines = content.lines();
    let mut name_opt: Option<&str> = None;

    // Find first non-empty line and non-comment line for the name
    for line in &mut lines {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        name_opt = Some(trimmed);
        break;
    }

    let name = name_opt.unwrap_or_else(|| {
        eprintln!("Error: no variable name specified");
        std::process::exit(1);
    });

    if opts.verbose {
        eprintln!("farchc: creating `{}`", name);
    }

    writeln!(output, "/* This file is generated by farchc. */").unwrap();
    writeln!(output).unwrap();
    writeln!(output, "#include <lib-common/farch.h>").unwrap();
    writeln!(output).unwrap();
    writeln!(output, "static const farch_data_t {}_data[] = {{", name).unwrap();

    let mut entries = Vec::<FarchEntry>::new();
    let mut lineno = 2;

    for line in lines {
        lineno += 1;

        if line.len() > PATHMAX as usize {
            eprintln!("Error: line {} is too long", lineno);
            std::process::exit(1);
        }

        let path = line.trim();
        if path.is_empty() {
            continue;
        }

        if path.starts_with('#') {
            if opts.verbose {
                eprintln!("farchc: {}", path);
            }
            continue;
        }

        let fullpath = reldir.join(path);

        if opts.verbose {
            eprintln!("farchc: adding `{}` as `{}`", path, fullpath.display());
        }

        let mut entry = FarchEntry {
            name: path.to_owned(),
            size: 0,
            compressed_size: 0,
            nb_chunks: 0,
        };

        writeln!(output, "/* {{{{{{ {} */", path).unwrap();

        dump_file(&fullpath, &mut entry, output);
        writeln!(output, "/* }}}}}} */").unwrap();

        entries.push(entry);
    }

    writeln!(output, "}};").unwrap();
    writeln!(output).unwrap();
    writeln!(output, "static const farch_entry_t {}[] = {{", name).unwrap();

    dump_entries(name, &entries, output);
    writeln!(output, "{{   .name = LSTR_NULL }},").unwrap();
    writeln!(output, "}};").unwrap();
}

fn main() {
    let opts = Opts::parse();

    // Handle input file
    let script_path = &opts.script;

    let reldir = script_path.parent().unwrap_or_else(|| {
        eprintln!(
            "Error: unable to get parent directory of script `{}`",
            script_path.display()
        );
        std::process::exit(1);
    });

    let mut output = String::new();

    do_work(&opts, reldir, &mut output);

    // Output to a file
    if let Some(out_path) = &opts.out {
        // Write file
        let mut out_file = File::create(out_path).unwrap_or_else(|e| {
            eprintln!(
                "Error: unable to open `{}` for writing: {}",
                out_path.display(),
                e
            );
            std::process::exit(1);
        });
        out_file.write_all(output.as_bytes()).unwrap_or_else(|e| {
            eprintln!(
                "Error: unable to write file `{}`: {}",
                out_path.display(),
                e
            );
            std::process::exit(1);
        });

        // Set permissions on output
        if let Err(e) = fs::set_permissions(out_path, fs::Permissions::from_mode(0o440)) {
            eprintln!("Error: unable to chmod `{}`: {}", out_path.display(), e);
            std::process::exit(1);
        }
    } else {
        // Output to stdout
        io::stdout()
            .write_all(output.as_bytes())
            .unwrap_or_else(|e| {
                eprintln!("Error: unable to write to stdio: {}", e);
                std::process::exit(1);
            });
    }
}
